superchain/superchain.go

```diff
@@ -61,6 +71,251 @@ type AddressList struct {
+}
+
+// AddressSet represents a set of addresses for a given
+// contract. They are keyed by the semantic version.
+type AddressSet map[string]Address
+
+// VersionedContract represents a contract that has a semantic version.
+type VersionedContract struct {
+	Version string  `json:"version"`
+	Address Address `json:"address"`
+}
+
+// Get will handle getting semantic versions from the set
+// in the case where the semver string is not prefixed with
+// a "v" as well as if it does have a "v" prefix.
+func (a AddressSet) Get(key string) Address {
+	if strings.HasPrefix(key, "v") {
+		if addr, ok := a[key]; ok {
+			return addr
+		}
+		if addr, ok := a[strings.TrimPrefix(key, "v")]; ok {
+			return addr
+		}
+	}
+	return a[key]
+}
+
+// Versions will return the list of semantic versions for a contract.
+// It handles the case where the versions are not prefixed with a "v".
+func (a AddressSet) Versions() []string {
+	keys := maps.Keys(a)
+	for i, k := range keys {
+		if !strings.HasPrefix(k, "v") {
+			keys[i] = "v" + k
+		}
+	}
+	semver.Sort(keys)
+	return keys
+}
+
+// Resolve will return a set of addresses that resolve a given
+// semantic version set.
+func (c ContractImplementations) Resolve(versions ContractVersions) (ImplementationList, error) {
+	var implementations ImplementationList
+	var err error
+	if implementations.L1CrossDomainMessenger, err = resolve(c.L1CrossDomainMessenger, versions.L1CrossDomainMessenger); err != nil {
+		return implementations, fmt.Errorf("L1CrossDomainMessenger: %w", err)
+	}
+	if implementations.L1ERC721Bridge, err = resolve(c.L1ERC721Bridge, versions.L1ERC721Bridge); err != nil {
+		return implementations, fmt.Errorf("L1ERC721Bridge: %w", err)
+	}
+	if implementations.L1StandardBridge, err = resolve(c.L1StandardBridge, versions.L1StandardBridge); err != nil {
+		return implementations, fmt.Errorf("L1StandardBridge: %w", err)
+	}
+	if implementations.L2OutputOracle, err = resolve(c.L2OutputOracle, versions.L2OutputOracle); err != nil {
+		return implementations, fmt.Errorf("L2OutputOracle: %w", err)
+	}
+	if implementations.OptimismMintableERC20Factory, err = resolve(c.OptimismMintableERC20Factory, versions.OptimismMintableERC20Factory); err != nil {
+		return implementations, fmt.Errorf("OptimismMintableERC20Factory: %w", err)
+	}
+	if implementations.OptimismPortal, err = resolve(c.OptimismPortal, versions.OptimismPortal); err != nil {
+		return implementations, fmt.Errorf("OptimismPortal: %w", err)
+	}
+	if implementations.SystemConfig, err = resolve(c.SystemConfig, versions.SystemConfig); err != nil {
+		return implementations, fmt.Errorf("SystemConfig: %w", err)
+	}
+	return implementations, nil
+}
+
+// resolve returns a VersionedContract that matches the passed in semver version
+// given a set of addresses.
+func resolve(set AddressSet, version string) (VersionedContract, error) {
+	if !strings.HasPrefix(version, "v") {
+		version = "v" + version
+	}
```

Comment:
If we define a semver type we could create a `Canonicalize()` method on it that automagically adds the prefix.

Reply:
(It would allow us to have this check be one lined)
```golang
version = version.Canonicalize()
```

Reply:
And if we really want a string param, can do
```golang
semver = Semver(version).Canonicalize()
```